### PR TITLE
feat(kvindexers): update method for binary pb indexer

### PR DESCRIPTION
**Pull request by florian-hoenicke**
Related ticket: 
https://jinaai.atlassian.net/browse/JINACORE-483

Adding update and delete operations for all key value indexers.

### CODE REVIEW

**Review by theUnkownName — DISMISSED** [review]
Looks good!

File: jina/executors/indexers/keyvalue.py
Comment (theUnkownName) [inline jina/executors/indexers/keyvalue.py:81]: Maybe just something more explicit on what `p, r, l `are?

Comment (hanxiao) [reply jina/executors/indexers/keyvalue.py:81]: fun fact: when I wrote something like this (often intentionally), it's usually a sign this code is hardcore and at low-level, to remind you Data Structure & Algorithms 101

**Review by hanxiao — CHANGES_REQUESTED** [review]
- this does not work, as discussed in https://jinaai.slack.com/archives/C018QC78L84/p1606823311131600?thread_ts=1606820889.125300&cid=C018QC78L84
- missing unit test

File: jina/executors/indexers/keyvalue.py
Comment (hanxiao) [inline jina/executors/indexers/keyvalue.py:90]: as far as i understand, this tries to write the updated values in the *same* position but with varied length, if the length is longer than the old one, then the next entry in the `body` is overflowed.

Comment (cristianmtr) [reply jina/executors/indexers/keyvalue.py:90]: Ah yes we're not getting the l, p, r associated with the last entry.

Comment (hanxiao) [reply jina/executors/indexers/keyvalue.py:90]: or is it simply appending, based on prev. `self._start`?

Comment (hanxiao) [reply jina/executors/indexers/keyvalue.py:90]: but then why isn't `self._start` updated though?

anyway, a unit test on visiting all entries after a couple of updates would be clear enough to reveal the problem

Comment (cristianmtr) [reply jina/executors/indexers/keyvalue.py]: Can you check the new version?

File: jina/executors/indexers/keyvalue.py
Comment (hanxiao) [inline jina/executors/indexers/keyvalue.py:41]: dont see the need of change here

File: jina/executors/indexers/keyvalue.py
Comment (hanxiao) [inline jina/executors/indexers/keyvalue.py:75]:
```suggestion

```

File: jina/executors/indexers/keyvalue.py
Comment (hanxiao) [inline jina/executors/indexers/keyvalue.py:90]: This is simply
```suggestion
            old_value = self.query(key)
```

File: jina/executors/indexers/keyvalue.py
Comment (hanxiao) [inline jina/executors/indexers/keyvalue.py:109]: isn't this self.add()?

Comment (cristianmtr) [reply jina/executors/indexers/keyvalue.py:109]: No, because we need the `adder` handler, and we need a custom `_start` position

File: jina/executors/indexers/keyvalue.py
Comment (cristianmtr) [reply jina/executors/indexers/keyvalue.py:75]: Yes, but it fails sometimes. Need to debug it more and figure out why

File: jina/executors/indexers/keyvalue.py
Comment (cristianmtr) [reply jina/executors/indexers/keyvalue.py]: Yep, was debugging something.

File: jina/executors/indexers/keyvalue.py
Comment (cristianmtr) [inline jina/executors/indexers/keyvalue.py:31]: @florian-hoenicke Not sure I understand what this is doing

Comment (cristianmtr) [reply jina/executors/indexers/keyvalue.py:31]: Checks if elements from `[1:]` in the first row are `(-1, ...)` ? What would that mean?

Comment (cristianmtr) [reply jina/executors/indexers/keyvalue.py:31]: Oh I see those are the markers of having been deleted. Can't you just delete the key from the header completely and then rewrite it? It's much smaller than the body so it shouldn't be a performance hindrance.

Or at least make that `(-1, ...)` a constant

Comment (cristianmtr) [inline jina/executors/indexers/keyvalue.py:31]: My comments got deleted somehow, writing again:

Cant you just rewrite the entire header? Much smaller than body, shouldnt be a perf hindrance.

Or at least make `(-1, ...)` a constant

File: tests/unit/executors/indexers/test_binary_indexer.py
Comment (cristianmtr) [inline tests/unit/executors/indexers/test_binary_indexer.py:73]: Shouldn't `.save(` be required here, as after `.add(`?

File: tests/unit/executors/indexers/test_binary_indexer.py
Comment (cristianmtr) [inline tests/unit/executors/indexers/test_binary_indexer.py:74]: `        assert idxer.size == 1` fails

File: jina/executors/indexers/keyvalue.py
Comment (florian-hoenicke) [inline jina/executors/indexers/keyvalue.py:82]: @cristianmtr what about the sharding case, where we want to have an update on all shards having the key. I think, if  a shard don't have a key, it should silently ignore an update or delete request. I don't think there is a way of knowing which key is persistent in which shard from outside.

**Comment by JoanFM** [comment]
Hey guys, fix the `commit-lint` problem asap (then rebasing becomes a mess).

File: tests/unit/executors/indexers/test_binary_indexer.py
Comment (cristianmtr) [reply tests/unit/executors/indexers/test_binary_indexer.py:99]: Yes, it passes, but I see your point

File: jina/executors/indexers/keyvalue.py
Comment (florian-hoenicke) [reply jina/executors/indexers/keyvalue.py]: Now, I created a constant for `(-1, -1, -1)`.

File: jina/executors/indexers/keyvalue.py
Comment (cristianmtr) [reply jina/executors/indexers/keyvalue.py:84]: I see what you mean. But isn't that something the tail pea should handle? It should catch these errors, right?

**Comment by cristianmtr** [comment]
> Hey guys, fix the `commit-lint` problem asap (otherwise rebasing becomes a mess).

Done

File: jina/executors/indexers/keyvalue.py
Comment (florian-hoenicke) [reply jina/executors/indexers/keyvalue.py:84]: okay, then let's leave it like this for now.

File: jina/executors/indexers/keyvalue.py
Comment (nan-wang) [inline jina/executors/indexers/keyvalue.py:76]: Is this part necessary?

Comment (JoanFM) [reply jina/executors/indexers/keyvalue.py:76]: I think this is not needed, but the problem is that the `test` is doing something that we are not doing. in the same context  managed indexer we are indexing, saving and querying

Comment (JoanFM) [reply jina/executors/indexers/keyvalue.py:76]: It is not necessary and dangerous for now. Query handler and Write handler I think should not be opened at the same time.

File: tests/unit/executors/indexers/test_binary_indexer.py
Comment (JoanFM) [inline tests/unit/executors/indexers/test_binary_indexer.py:49]: this test should be changed:

1. Using `test_metas` fixture will give what u expect
2. What is 'pbidx'
3. For now, I think we cannot `add` and `query` in the same `context managed` indexer. You should close the context after the addition, and then open a new indexer for querying, and then yet another for updating, and yet another for later query.  (I think this holds true specially for VectorIndexers and I think also for BinaryPbIndexer)

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: jina/executors/indexers/keyvalue.py
Comment (JoanFM) [reply jina/executors/indexers/keyvalue.py:76]: I think the reason refers to the comments I added in the test

Comment (cristianmtr) [reply jina/executors/indexers/keyvalue.py]: It was removed, but it a symptom of another problem we discussed over zoom. basically we need to flip the mutex and close handlers between ops. This is now done in the `.update`

**Comment by florian-hoenicke** [comment]
@JoanFM @hanxiao Is there anything left to do for this pull request?

File: tests/unit/executors/indexers/test_binary_indexer.py
Comment (JoanFM) [inline tests/unit/executors/indexers/test_binary_indexer.py:35]: use test_metas fixture in our conftest.py

Comment (JoanFM) [reply tests/unit/executors/indexers/test_binary_indexer.py:35]: This test is too long. Could you divide it in different ones? I see comments that suggest that different behaviors are being tested and I think is better if they are reflected in a separate test

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: jina/executors/indexers/keyvalue.py
Comment (bhavsarpratik) [inline jina/executors/indexers/keyvalue.py:88]: Anything left to fix here?

File: jina/executors/indexers/keyvalue.py
Comment (bhavsarpratik) [inline jina/executors/indexers/keyvalue.py:71]: Remove the comment

File: tests/unit/executors/indexers/test_binary_indexer.py
Comment (bhavsarpratik) [inline tests/unit/executors/indexers/test_binary_indexer.py:37]: This can be a pytest fixture used for below tests broken into many tests.

**Review by bhavsarpratik — CHANGES_REQUESTED** [review]

File: jina/executors/indexers/keyvalue.py
Comment (cristianmtr) [reply jina/executors/indexers/keyvalue.py:88]: Wasn't sure if there is a way to get the name of the indexer. What i'm logging here is the path to the file instead

File: jina/executors/indexers/keyvalue.py
Comment (cristianmtr) [reply jina/executors/indexers/keyvalue.py]: done

File: tests/unit/executors/indexers/test_binary_indexer.py
Comment (cristianmtr) [reply tests/unit/executors/indexers/test_binary_indexer.py]: sure

File: tests/unit/executors/indexers/test_binary_indexer.py
Comment (cristianmtr) [reply tests/unit/executors/indexers/test_binary_indexer.py]: yes

File: tests/unit/executors/indexers/test_binary_indexer.py
Comment (bhavsarpratik) [inline tests/unit/executors/indexers/test_binary_indexer.py:81]: Remove this if not needed

**Review by bhavsarpratik — APPROVED** [review]

File: jina/executors/indexers/keyvalue.py
Comment (bwanglzu) [inline jina/executors/indexers/keyvalue.py:87]: I guess you don't need to loop over all the keys and raise the error.
```python
for key in keys:
    if not self.query_handler.header.get(key):
        raise KeyError('...')
```

File: jina/executors/indexers/__init__.py
Comment (bwanglzu) [inline jina/executors/indexers/__init__.py:222]: curious, why not `__setitem__` and `__del__`?

File: tests/unit/executors/indexers/test_binary_indexer.py
Comment (bwanglzu) [inline tests/unit/executors/indexers/test_binary_indexer.py:74]: in `test_metas` we already set `tmpdir`, so you don't have to use the `tmpdir` fixture anymore

**Review by bwanglzu — CHANGES_REQUESTED** [review]

File: jina/executors/indexers/__init__.py
Comment (florian-hoenicke) [reply jina/executors/indexers/__init__.py:222]: Nice catch! Actually, I don't see any argument for not using them. Changing this could be done in another ticket.
@maximilianwerk Do you know why we don't use dunder methods here?

File: jina/executors/indexers/keyvalue.py
Comment (florian-hoenicke) [reply jina/executors/indexers/keyvalue.py:87]: Nice point. True, all missed update-keys are collected in memory. As a tradeoff we get have a better error handling. 
Actually, we have all keys in memory already (ReadHandler.header). Since keys are expected to be much smaller than values, we decided to accept the tradeoff.

File: tests/unit/executors/indexers/test_binary_indexer.py
Comment (florian-hoenicke) [reply tests/unit/executors/indexers/test_binary_indexer.py]: thanks, I removed the tmpdir parameter

**Review by cristianmtr — APPROVED** [review]
LGTM now. Any other feedback? We should merge this

File: jina/executors/indexers/keyvalue.py
Comment (cristianmtr) [reply jina/executors/indexers/keyvalue.py:87]: We want to log them all in one go

**Review by JoanFM — CHANGES_REQUESTED** [review]
We cannot merge this, you have done some updates to the `hub` submoudle that should not be there.

**Comment by cristianmtr** [comment]
> We cannot merge this, you have done some updates to the `hub` submoudle that should not be there.

Not sure how that happened, but removed. Should I squash the commits before merging?

**Review by JoanFM — APPROVED** [review]

File: jina/executors/indexers/__init__.py
Comment (maximilianwerk) [reply jina/executors/indexers/__init__.py:222]: Currently, `__getitem__` is just calling `query` in the `BaseKVIndexer`. Following this approach, we would need to implement `__setitem__` with either calling `add` or `update`. Since we bind the CRUD operations even to dedicated request objects, it would be weird to first make an explicit update call, and then convert it to a `__setitem__` which needs to figure out again, whether it is a `add` or `update`. Anyhow, for convenience, we should add the `__del__` and `__setitem__` later on.

File: jina/executors/indexers/keyvalue.py
Comment (maximilianwerk) [inline jina/executors/indexers/keyvalue.py:84]: Can you please add an integration test for two updates calls in a row on the same `idxer` object? I'd expect this line to fail then. If that is intended, then write a test, which expects to fail in the second `update` call in order to document, that double update on the very same `idxer` object is not possible.

File: tests/unit/executors/indexers/test_binary_indexer.py
Comment (bwanglzu) [reply tests/unit/executors/indexers/test_binary_indexer.py]: @florian-hoenicke great! while I see another two , can you also remove them?

**Review by bwanglzu — DISMISSED** [review]
lgtm if you can remove the unused tmpdirs.

File: jina/executors/indexers/keyvalue.py
Comment (cristianmtr) [reply jina/executors/indexers/keyvalue.py:84]: You mean like this https://github.com/jina-ai/jina/blob/e318088a30897ba3ca538de8e0f5f471af92688c/tests/unit/executors/indexers/test_binary_indexer.py#L119 ? Seems to pass

**Comment by cristianmtr** [comment]
> lgtm if you can remove the unused tmpdirs.

Done, removed

**Review by bwanglzu — APPROVED** [review]
LGTM!

File: jina/executors/indexers/keyvalue.py
Comment (maximilianwerk) [reply jina/executors/indexers/keyvalue.py:84]: Yes exactly. Even better.

File: jina/executors/indexers/keyvalue.py
Comment (JoanFM) [inline jina/executors/indexers/keyvalue.py:71]: shoul we maybe have a lazy flush? and flush only if we need the guarantee? or it feels like early optimization?

File: tests/unit/executors/indexers/test_binary_indexer.py
Comment (JoanFM) [inline tests/unit/executors/indexers/test_binary_indexer.py:78]: this may be fixed with the lazy flush?

Comment (JoanFM) [reply tests/unit/executors/indexers/test_binary_indexer.py:78]: I think add and update should be supported; if we merge this please create a ticket to keep track of this

Comment (cristianmtr) [reply tests/unit/executors/indexers/test_binary_indexer.py:78]: Yes, I will

Comment (cristianmtr) [reply tests/unit/executors/indexers/test_binary_indexer.py:78]: https://github.com/jina-ai/jina/issues/1413

File: jina/executors/indexers/keyvalue.py
Comment (cristianmtr) [reply jina/executors/indexers/keyvalue.py:71]: Opened an issue https://github.com/jina-ai/jina/issues/1413

**Pull request merged**
